### ISSUE TITLE
Uncaught unconditional recursion in `From` impl

### ISSUE THREAD

**Issue by CyrusAlbright** [rust-lang/rust#98217]
Rustc has no issues compiling the following code ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=016b8aea3b49c0fb008bb35aa162c74c)):
```rust
#![deny(unconditional_recursion)]

struct Dummy;

impl From<u8> for Dummy {
    fn from(u: u8) -> Self {
        u.into()
    }
}
```

The other way of writing this, however, does (correctly) fail to compile ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=503d9f7de7c1dee0e452ed896a1311ff)):
```rust
#![deny(unconditional_recursion)]

struct Dummy;

impl From<u8> for Dummy {
    fn from(u: u8) -> Self {
        Dummy::from(u)
    }
}
```

I assume that type inference is happening after the `unconditional_recursion` lint causing it to miss this?

**Comment by eggyal** [comment]
The lint only detects direct recursion (where a function calls itself).  It doesn't detect recursion through other functions, so the following will compile ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=40d5142263a81555c302d222a4c224c4)):

```rust
#![deny(unconditional_recursion)]

pub fn a() {
    b()
}

pub fn b() {
    a()
}
```

In your example, `<Dummy as From<u8>>::from` is calling `<u8 as Into<Dummy>>::into` which then calls back to the former.

**Comment by jruderman** [comment]
This is a case of #57965 where the recursion goes through trait functions.
